helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,518 @@
+use std::borrow::Cow;
+
+use anyhow::{anyhow, Result};
+use helix_core::SmallVec;
+
+use crate::{util::lsp_pos_to_pos, OffsetEncoding};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum CaseChange {
+    Upcase,
+    Downcase,
+    Capitalize,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum FormatItem<'a> {
+    Text(&'a str),
+    Capture(usize),
+    CaseChange(usize, CaseChange),
+    Conditional(usize, Option<&'a str>, Option<&'a str>),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Regex<'a> {
+    value: &'a str,
+    replacement: Vec<FormatItem<'a>>,
+    options: Option<&'a str>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum SnippetElement<'a> {
+    Tabstop {
+        tabstop: usize,
+    },
+    Placeholder {
+        tabstop: usize,
+        value: Box<SnippetElement<'a>>,
+    },
+    Choice {
+        tabstop: usize,
+        choices: Vec<&'a str>,
+    },
+    Variable {
+        name: &'a str,
+        default: Option<&'a str>,
+        regex: Option<Regex<'a>>,
+    },
+    Text(&'a str),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Snippet<'a> {
+    elements: Vec<SnippetElement<'a>>,
+}
+
+pub fn parse(s: &str) -> Result<Snippet<'_>> {
+    parser::parse(s).map_err(|rest| anyhow!("Failed to parse snippet. Remaining input: {}", rest))
+}
+
+pub fn into_transaction<'a>(
+    snippet: Snippet<'a>,
+    doc: &helix_core::Rope,
+    selection: &helix_core::Selection,
+    edit: &lsp_types::TextEdit,
+    line_ending: &str,
+    offset_encoding: OffsetEncoding,
+    include_placeholer: bool,
+) -> helix_core::Transaction {
+    use helix_core::{smallvec, Range, Transaction};
+    use SnippetElement::*;
+
+    let text = doc.slice(..);
+    let primary_cursor = selection.primary().cursor(text);
+
+    let start_offset = match lsp_pos_to_pos(doc, edit.range.start, offset_encoding) {
+        Some(start) => start as i128 - primary_cursor as i128,
+        None => return Transaction::new(doc),
+    };
+    let end_offset = match lsp_pos_to_pos(doc, edit.range.end, offset_encoding) {
+        Some(end) => end as i128 - primary_cursor as i128,
+        None => return Transaction::new(doc),
+    };
+
+    let newline_with_offset = format!(
+        "{line_ending}{blank:width$}",
+        width = edit.range.start.character as usize,
+        blank = ""
+    );
```

Comment:
I think you'll need to compute this differently for each cursor since the offset may be different for each of them.
It may be better to perform snippet -> (Text, tabstop) translation for each selection independently since any rule that depends on the cursor location (newlines and eventually variables will require that eventually).

Reply:
I took a stab at such refactoring - https://github.com/Urgau/helix/pull/2 . It does fix the issue with multiple cursors at different offsets